File: epi_judge_java/epi/SubstringMatch.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

public class SubstringMatch {
    private static final int PRIME = 37;

    @EpiTest(testDataFile = "substring_match.tsv")
    // Returns the index of the first character of the substring if found, -1
    // otherwise.
    public static int rabinKarp(String t, String s) {
        if (t.length() < s.length()) return -1;
        long pow = 1;
        long tHash = 0;
        long sHash = 0;

        int n = s.length();

        for (int i = 0; i < n; i++) {
            tHash = tHash * PRIME + t.charAt(i);
            sHash = sHash * PRIME + s.charAt(i);

            if (i != 0) {
                pow *= PRIME;
            }
        }

        if (sHash == tHash && eq(t, 0, s)) {
            return 0;
        }

        for (int i = 1; i <= t.length() - n; i++) {
            tHash -= pow * t.charAt(i - 1);
            tHash = tHash * PRIME + t.charAt(i + n - 1);

            if (sHash == tHash && eq(t, i, s)) {
                return i;
            }
        }

        return -1;
    }

    private static boolean eq(String t, int left, String s) {
        for (int i = 0; i < s.length(); i++) {
            if (t.charAt(i + left) != s.charAt(i)) return false;
        }

        return true;
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "SubstringMatch.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/substring_match.cc
#include <string>

#include "test_framework/generic_test.h"

using std::string;

bool eq(const string &t, const string &s, int i);

// Returns the index of the first character of the substring if found, -1
// otherwise.
int RabinKarp(const string &t, const string &s) {
    if (s.size() > t.size()) return -1;

    int prime = 37;
    long long tHash = 0;
    long long sHash = 0;
    long long mult = 1;

    for (int i = 0; i < s.size(); ++i) {
        tHash = tHash * prime + t[i];
        sHash = sHash * prime + s[i];

        if (i != 0) {
            mult *= prime;
        }
    }

    if (tHash == sHash) {
        if (eq(t, s, 0)) {
            return 0;
        }
    }

    for (int i = 1; i + s.size() <= t.size(); ++i) {
        tHash -= t[i - 1] * mult;
        tHash *= prime;
        tHash += t[i + s.size() - 1];

        if (tHash == sHash && eq(t, s, i)) {
            return i;
        }
    }

    return -1;
}

bool eq(const string &t, const string &s, int start) {
    for (int i = 0; i < s.size(); ++i) {
        if (t[i + start] != s[i]) return false;
    }

    return true;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"t", "s"};
    return GenericTestMain(args, "substring_match.cc", "substring_match.tsv",
                           &RabinKarp, DefaultComparator{}, param_names);
}
